File: medium/2352. Equal Row and Column Pairs/2352. Equal Row and Column Pairs.go
package main

import (
	"fmt"
	"strings"
)

func equalPairs(grid [][]int) int {
	row := make(map[string]int)
	n := len(grid)

	for r := 0; r < n; r++ {
		key := encode(grid[r])
		row[key] += 1
	}

	res := 0

	for c := 0; c < n; c++ {
		col := make([]int, n)

		for r := 0; r < n; r++ {
			col[r] = grid[r][c]
		}

		key := encode(col)
		res += row[key]
	}

	return res
}

func encode(arr []int) string {
	sb := strings.Builder{}
	for _, num := range arr {
		sb.WriteString(fmt.Sprintf("%d,", num))
	}
	return sb.String()
}

File: medium/2352. Equal Row and Column Pairs/2352. Equal Row and Column Pairs.py
from collections import defaultdict
from typing import List


class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        # ver: defaultdict[tuple, int] = defaultdict(int)
        # hor: defaultdict[tuple, int] = defaultdict(int)

        # n = len(grid)

        # for r in range(n):
        #     hor[tuple(grid[r])] += 1
        #     rev = tuple(reversed(grid[r]))
        #     hor[rev] += 1

        # for c in range(n):
        #     key = []
        #     for r in range(n):
        #         key.append(grid[r][c])

        #     ver[tuple(key)] += 1
        #     rev = tuple(reversed(key))
        #     ver[rev] += 1

        # res = 0

        # for key in ver:
        #     while ver[key]:
        #         if key in hor and hor[key]:
        #             res += 1
        #             hor[key] -= 1

        #         ver[key] -= 1

        # return res

        row = defaultdict(int)
        n = len(grid)

        for r in range(n):
            row[tuple(grid[r])] += 1

        res = 0

        for c in range(n):
            col = tuple(grid[r][c] for r in range(n))
            res += row[col]

        return res


solution = Solution()
print(solution.equalPairs(grid=[[3, 2, 1], [1, 7, 6], [2, 7, 7]]))
